packages/cloud_firestore/cloud_firestore_web/lib/utils/codec_utility.dart

```diff
@@ -0,0 +1,95 @@
+part of cloud_firestore_web;
+
+// disabling lint as it's only visible for testing
+@visibleForTesting
+// ignore: public_member_api_docs
+class CodecUtility {
+  // disabling lint as it's only visible for testing
+  @visibleForTesting
+  // ignore: public_member_api_docs
+  static Map<String, dynamic> encodeMapData(Map<String, dynamic> data) {
+    if (data == null) {
+      return null;
+    }
+    Map<String, dynamic> output = Map.from(data);
+    output.updateAll((key, value) => valueEncode(value));
+    return output;
+  }
+
+  // disabling lint as it's only visible for testing
+  @visibleForTesting
+  // ignore: public_member_api_docs
+  static List<dynamic> encodeArrayData(List<dynamic> data) {
+    if (data == null) {
+      return null;
+    }
+    return List.from(data).map(valueEncode).toList();
+  }
+
+  // disabling lint as it's only visible for testing
+  @visibleForTesting
+  // ignore: public_member_api_docs
+  static dynamic valueEncode(dynamic value) {
+    if (value is FieldValueInterface && value.instance is FieldValueWeb) {
+      return (value.instance as FieldValueWeb)._delegate;
+    } else if (value is Timestamp) {
+      return value.toDate();
+    } else if (value is GeoPoint) {
+      return web.GeoPoint(value.latitude, value.longitude);
+    } else if (value is Blob) {
+      return web.Blob.fromUint8Array(value.bytes);
+    } else if (value is DocumentReferenceWeb) {
+      return value.delegate;
+    } else if (value is Map<String, dynamic>) {
+      return encodeMapData(value);
+    } else if (value is List<dynamic>) {
+      return encodeArrayData(value);
+    }
+    return value;
+  }
+
+  // disabling lint as it's only visible for testing
+  @visibleForTesting
+  // ignore: public_member_api_docs
+  static Map<String, dynamic> decodeMapData(Map<String, dynamic> data) {
+    if (data == null) {
+      return null;
+    }
+    Map<String, dynamic> output = Map.from(data);
+    output.updateAll((key, value) => valueDecode(value));
+    return output;
+  }
+
+  // disabling lint as it's only visible for testing
+  @visibleForTesting
+  // ignore: public_member_api_docs
+  static List<dynamic> decodeArrayData(List<dynamic> data) {
+    if (data == null) {
+      return null;
+    }
+    return List.from(data).map(valueDecode).toList();
+  }
+
+  // disabling lint as it's only visible for testing
+  @visibleForTesting
+  // ignore: public_member_api_docs
+  static dynamic valueDecode(dynamic value) {
+    if (value is web.GeoPoint) {
+      return GeoPoint(value.latitude, value.longitude);
+    } else if (value is DateTime) {
+      return Timestamp.fromDate(value);
```

Comment:
Hm... wouldn't this cause crashes if I try to save a structure like this to a map type in firestore:

`Map<String, List<DateTime>>`

The next time I load it, it'll be cast to:

`Map<String, List<Timestamp>>`

And we'll just crash at a different point.

I think that with the current backend library, we cannot reliably do this Timestamp <-> DateTime conversion, because we lose the information of what the user was *actually attempting to store*: a Timestamp or a DateTime?

Reply:
True and that's also true for `cloud_firestore: ^0.13.0+1` on mobile. The only upside for this is that it makes mobile and web consistent in that in only outputs Timestamp.

Reply:
> True and that's also true for `cloud_firestore: ^0.13.0+1` on mobile. The only upside for this is that it makes mobile and web consistent in that in only outputs Timestamp.

:epic_facepalm:

Reply:
Is it documented anywhere that Timestamps are the only "safe" way of handling date/time in cloud_firestore?

Reply:
Not sure if it's documented. I tested it using the example project and that's how I found this behaviour.

Reply:
Well, if this makes web (mis)behave consistently with the mobile implementation, I guess it's all good!

I'm going to update my app to the latest version of this branch and keep going.

Other than this, and some self-inflicted errors for not cancelling Stream listeners when needed, this is solid!

Reply:
Awesome.. Thanks. I am sorting out an issue with handling the FirebaseApp instance, just avoiding to recreate it unnecessarily. I will push this change soon.

Reply:
Hey, author of #46 here. What exactly *is* the practical difference between a `DateTime` and a `Timestamp`?